packages/apollo-server-defer/README.md

```diff
@@ -0,0 +1,173 @@
+
+- _Expensive to load_. This includes private data that is not cached (like user progress), or information that requires more computation on the backend (like calculating price quotes on Airbnb)
+- _Not on the critical path for interactivity._ This includes loading the comments section of a blog post, or the number of claps received.
+- _Expensive to send._ Even if the field may resolve quickly (ready in memory), users might still choose to defer it if the cost of transport is too expensive.
+
+#### Use Cases
+
+```graphql
+type Video {
+  # public data that can be cached
+  id: ID!
+  title: String
+  description: String
+  # progress is private to the user, and significantly more
+  # expensive to retrieve
+  progress: Progress
+  comments: [Comment]
+}
+
+type Progress {
+  percent: Float
+  position: Float
+}
+
+type Comment {
+  userId: ID!
+  text: String
+}
+
+# Example Query using @defer
+query {
+  video(id: $id) {
+    title: String
+    progress @defer {
+      percent
+      position
+    }
+    comments @defer {
+      userId
+      text
+    }
+  }
+}
+```
+
+Instead of holding the GraphQL response and page rendering until the entire query is resolved, @defer tells Apollo Server to return a partial query response if the deferred fields are not ready yet. In the example above, we see how it may be used to remove `progress` and `comments` from the intial query response. Apollo Server would then take care of resolving the rest of the deferred fields in the background, and stream them to the client when they are ready.
+
+```
+// Initial Response
+{
+  "asset": {
+    "title": "Westworld",
+  }
+}
+
+// Stream in deferred fields as patches
+{
+  "path": ["video", "progress"],
+  "data": {
+    "percent": 20,
+    "position": 20
+  }
+}
+{
+  "path": ["video", "comments"],
+  "data": [{
+    "userId": "123",
+    "text": "..."
+  }]
+}
+```
+
+#### Who needs this the most?
+
+- Media companies, banks, other data-heavy enterprise applications.
+
+#### Why is @defer an ideal solution?
+
+- Query remains super easy to read 👍
+- Stays true to Apollo's declarative approach to data loading - everything is encapsulated in a single query component 😎
+- Greatly reduces boilerplate code to manage multiple requests, a huge value-add for users that use both Apollo Server AND Client together.
+
+### Implementation Details
+
+#### Patch Data Format:
+
+- Proposed format for `ExecutionPatchResult`
+  ```graphql
+  {
+    path: [string]!
+    data: {}
+    errors: []
+  }
+  ```
+- Current transport assumes ExecutionResult so some changes need to be made.
+- We should use the same patch format for @defer/stream/live support.
+
+#### Apollo Server:
+
+In order to support @defer, there are significant changes to be made to the execution phase if GraphQL.
+
+- Roll our own derivative of graphql.js execution that supports deferred responses and streaming patches using observables.
```

Comment:
I would characterize this as **extending** graphql-js execution to support deferred responses. I think its important for us to compare the value of async itterable (which already has support for subscriptions) and observables. I like observables but want to have strong reason for adding another data type within the system.

However, since we don't **expose** this data type, it may not even be worth mentioning and instead keep it as a changeable implementation detail. I do think that we may one day want to allow returning of an async iterable or observable for control of something like `stream / live` so lets give this some careful thought.

@evans I know you have done some thinking here, care to chime in?